Missing data summary, categorical column bar plots

cbharp opened the pull request:
- Added counts and percentages of missing values for minimal dataset 
- Added bar charts showing percentages for categorical values columns of minimal dataset

@cbharp Please update the title/ description to reflect the list of changes made in this PR

korelasidata reviewed (changes requested):
Thanks for working on the bar charts. They look great. I'm already noticing a few insights I'd like to flag to the client. Great work. Can you include the name of the data source in the plot title so we know where to find each variable plotted? For example (making this up): "HMDA: derived_ethnicity". Otherwise, I think we'll be ready to merge.

korelasidata commented on banking_dashboard/run-data-ingestion.ipynb:
In `hmda_dict['lar_df']`, filter down to the 3 counties of interest. You can leave the`hmda_dict['panel_df']` as-is.

<img width="541" alt="image" src="https://github.com/KorelasiDataInsights/southern_dallas_progress/assets/76417140/b9adf10e-3c11-4e8a-94c5-9cc6d95c3cbb">

korelasidata replied:
General comment: Strip the whitespace out of column names in all datasets. For example, in `fdic_locations_df`, rename `Branch State   ` to `Branch State`

korelasidata reviewed (changes requested):
Nice work.  Please note the comments attached to this review.

korelasidata commented:
In the Jupyter notebook, add each of the following groups of value counts in the chunk immediately after each relevant data source is created. For example, insert the first set after `census_df` is created.

```
print(census_df['state'].value_counts())
print(census_df['county'].value_counts())
print(census_df['tract'].value_counts())

print(ffiec_data['FIPS state code'].value_counts())
print(ffiec_data['FIPS county code'].value_counts())
print(ffiec_data['Census tract. Implied decimal point'].value_counts())

print(hmda_dict['lar_df']['state_code'].value_counts())
print(hmda_dict['lar_df']['county_code'].value_counts())
print(hmda_dict['lar_df']['census_tract'].value_counts())

print(final_cra_dict['cra2021_Discl_D11.dat']['State'].value_counts())
print(final_cra_dict['cra2021_Discl_D11.dat']['County'].value_counts())

print(final_cra_dict['cra2021_Discl_D6.dat']['State'].value_counts())
print(final_cra_dict['cra2021_Discl_D6.dat']['County'].value_counts())
print(final_cra_dict['cra2021_Discl_D6.dat']['Census Tract'].value_counts())

print(fdic_locations_df['Branch State   '].value_counts())
print(fdic_locations_df['Branch County'].value_counts())
# print(fdic_locations_df['Census tract'].value_counts())

print(sba_data['Borrower state'].value_counts())
#print(sba_data['Branch County'].value_counts())
# print(sba_data['Census tract'].value_counts())
```

korelasidata commented:
@cbharp Filter SBA data down to Borrower state = Texas.

korelasidata commented:
@cbharp Use this logic for identifying the county associated with each borrower address in the SBA dataset. You will need to install the `zipcodes` package. 
```
import zipcodes
import pandas as pd

counties_of_interest = ["Dallas County", "Collin County", "Tarrant County"]

def get_zip_codes(county_name, abbrev_state = "TX"):
  if " County" not in county_name:
     county_name = county_name + " County"
  
  zip_list = zipcodes.filter_by(county = county_name, state = abbrev_state)

  zips_raw = [i['zip_code'] for i in zip_list]
  zips_unique = pd.Series(zips_raw).drop_duplicates().tolist()
  zips_unique.sort()

  print(f"{len(zips_unique)} unique zip codes found in {county_name}")

  return zips_unique
  

zip_dallas = get_zip_codes(county_name = "Dallas County", abbrev_state = "TX")
zip_collin = get_zip_codes(county_name = "Collin County", abbrev_state = "TX")
zip_tarrant = get_zip_codes(county_name = "Tarrant County", abbrev_state = "TX")

all_zips = zip_dallas + zip_collin + zip_tarrant
print(len(all_zips))
```

After filtering down to Texas (to make the dataset smaller and the mapping faster), create a new field called `County` . You can create a mapper that checks whether it's in each of the 3 counties of interest and maps it accordingly. If it isn't in any of these counties, just code it as "Other".

korelasidata commented:
@cbharp  Use the following code to map address to census tract for (1) FDIC Locations and (2) SBA


```

# Install this module. Don't forget to update `requirements.txt`
# !pip install census-geocoder

# Load the module
import census_geocoder as geocoder

# Helper function
# TO DO: 
# - Add type hinting
# - Add doc string
# - Add error handling (try/except)... if exception occurs, return dict with same structure as normal but populate values with None
# - Convert census tract to standard format (use helper function)

# address_str: single line address, see examples below

def get_census_geocode(address_str):
  
  # Get census geocoding
  geo_dict = geocoder.geography.from_address(address_str).__dict__

  # Get geographical details
  geo = geo_dict['extensions']['result']['addressMatches'][0]['geographies']

  # Get census tract(s)... hopefully there's only one?
  # If more than 1, print a warning: print(f"{len(tracts)} zip codes found. Using first result")
  # Similar for states and counties

  census_tracts = [i['TRACT'] for i in geo['Census Tracts']]
  counties = [i['BASENAME'] for i in geo['Counties']]
  states = [i['BASENAME'] for i in geo['States']]


  return {
      'state': states[0],
      'county': counties[0],
      'census_tract': census_tracts[0],
  }
```

Test results

```
# Test 1
get_census_geocode("1127 G Ave, Plano, TX 75074")

# Result 1
# {'state': 'Texas', 'county': 'Collin', 'census_tract': '031902'}

# Test 2
get_census_geocode("1117 Montclair Dr, Garland, TX 75040")

# Result 2
{'state': 'Texas', 'county': 'Dallas', 'census_tract': '018203'}

# Test 3
get_census_geocode("5661 Comer Dr, Fort Worth, TX 76134")

# Result 3
{'state': 'Texas', 'county': 'Tarrant', 'census_tract': '106002'}
```

Use this logic to create new county and census tract fields (with a different county name for this logic compared to the one you created using the zip code logic above). Let's see if the county results in this logic match the zipcode-based logic above. You may need to standardize the county name so they both end in " County"

korelasidata commented:
@cbharp The new census data ingestor looks good. Let the new version replace the old one (i.e., delete the old one and use the old name for the new function)

cbharp commented:
> @cbharp The new census data ingestor looks good. Let the new version replace the old one (i.e., delete the old one and use the old name for the new function). And of course update the function call in the Jupyter notebook accordingly.

Reply:
@korelasidata 
I have the following completed:

General
- added value_counts() blocks after each data source is read in 

FDIC:
- locations data:
	-> added a "County_from_Zipcode" field based on the zipcode of the bank branch and using the "get_zip_codes" function you wrote. I think naming the new field "Branch_County_from_Zipcode" instead might prevent potential confusion. 

	-> added "county_from_address"and "census_tract_from_address" fields based on the branch address using the "get_census_geocode" function you wrote. 
	**NOTE** The function takes 12 minutes to run on ~1285 rows so I am trying to make changes so it runs faster.  

SBA Data:
- filtered for entires that are in Texas only(also added in full state name)

- created new column called "County_from_Zipcode" for county associated with borrower zip code. Would it be helpful to rename this column "Borrower_County_from_Zipcode"?

- adding in county and census tract fields based on SBA data address
-
- **NOTE** The function takes ~42 minutes to run on ~3800 rows.

korelasidata commented:
@cbharp Sharing the code we reviewed today re: census geocoding in batch mode

Note:   Input CSV file is expected to have 5 columns:
*   Key/identifier (in the SBA file, I'm using the business name)
*   Street address
*   City
*   State
*   Zip code


I believe the Census Geocode API does some pre-filtering to make sure that all rows are unique. So if there are any duplicate records, the result will contain fewer rows, and there won't be a one-to-one match between the input and output. 


TO DO:
*   Filter down SBA data to Texas first before applying the function (I believe this is already done on your end)
*   Run in batches of up to 10k records, as the API can only accept up to 10k at a time (likely not relevant given the small row counts, so you can IGNORE this part for now)
*   Adjust the function so that it returns the same number of rows as in the input file. Must split on the single-line-address field so that the address components (street address, city, state, zip) are all represented as separate columns




```
# !pip install census-geocoder
# !pip install validators

import census_geocoder as geocoder
import pandas as pd

pd.options.display.max_rows = None
pd.options.display.max_columns = None

import os
from validator_collection import validators
from census_geocoder.constants import CENSUS_API_URL
from backoff_utils import backoff
import requests
import csv

# Formatting the input 
sba = pd.read_csv('sba_data_borrower_addresses.csv') # , nrows=10000
sba = sba[sba["Borrower state"] == "TX"]
del sba["Unnamed: 0"]


# Helper function
def census_batch_lookup(filename='sba_sample.csv'):

  DEFAULT_BENCHMARK = os.environ.get('CENSUS_GEOCODER_BENCHMARK', 'CURRENT')
  DEFAULT_VINTAGE = os.environ.get('CENSUS_GEOCODER_VINTAGE', 'CURRENT')
  DEFAULT_LAYERS = os.environ.get('CENSUS_GEOCODER_LAYERS', 'all')

  print("Validate file existence")
  file_ = validators.file_exists(filename, allow_empty = False)

  print("Get batch addresses")
  result_ = geocoder.geography._get_batch_addresses(file_ = file_,
                                          benchmark = DEFAULT_BENCHMARK,
                                          vintage = DEFAULT_VINTAGE,
                                          layers = DEFAULT_LAYERS)
  
  print("Get results that were non-failures")
  res_nonfail = [i for i in result_ if len(i) >3]
  print(f"{len(res_nonfail)} out of {len(result_)} succeeded")

  print("Extract geographic components")
  res2 = [geocoder.geography.from_csv_record(x) for x in res_nonfail]

  print("Extract components as dicts")
  out = [i.__dict__ for i in res2]

  print("Format result")
  out_df = pd.DataFrame(out)
  out_df['_latitude'] = [out_df['extensions'][k]['latitute'] for k in range(0,out_df.shape[0])]
  out_df.dropna(how='all', axis=1, inplace=True) 
  out_df['Borrower name'] = [i[0] for i in res_nonfail]
  out_df['Borrower street address'] = [i[1] for i in res_nonfail]
  out_df['Match result'] = [i[2] for i in res_nonfail]
  out_df['Match type'] = [i[3] for i in res_nonfail]

  print("Done")

  return out_df

# Run
b = census_batch_lookup()
b.head()

```

korelasidata commented:
@cbharp  Notes from today's call

- [ ] **SBA data** - map county FIPS codes to county names
- [ ] **FDIC locations data** - apply both zip code and census geocode logic
- [ ] **FDIC locations data** - use groupby statements to prepare aggregate summaries (counts) of records for COUNTY only. We want to know how often the `Branch county` field in the original dataset agrees/disagrees with the zipcode county and census geocode county. No need to validate the other fields.

korelasidata commented:
@cbharp 

- [ ] 1. You have computed county based on two methods. They may agree or disagree, and we haven't yet made any determination about which is the "final" county. So I'm asking you to create a new field that essentially coalesces the two county computations. If census geocode county is available, use that one. If it is NULL, then use the zip code based county. Maybe call the new field 'County final". Just for SBA... no need to do this for FDIC locations
- [ ] 2. Filter down to Collin, Dallas, and Tarrant counties in the SBA dataset based on the "County final" field
- [ ] 3. Please make sure that a preview of the SBA dataset (the first 100 rows) is visible in the Jupyter notebook that you commit to the repo.
